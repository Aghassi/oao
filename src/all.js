// @flow

import { removeAllListeners, addListener } from 'storyboard';
import listPaths from './utils/listPaths';
import { exec } from './utils/shell';
import parallelConsoleListener from './utils/parallelConsoleListener';

type Options = {
  src: string,
  parallel?: boolean,
  parallelLogs?: boolean,
  ignoreErrors?: boolean,
};

const run = async (cmd: string, {
  src: srcPatterns,
  parallel,
  parallelLogs,
  ignoreErrors,
}: Options) => {
  if (parallel && parallelLogs) {
    removeAllListeners();
    addListener(parallelConsoleListener);
  }
  const pkgPaths = await listPaths(srcPatterns);
  const allPromises = [];
  for (let i = 0; i < pkgPaths.length; i += 1) {
    let promise = exec(cmd, { cwd: pkgPaths[i], bareLogs: parallelLogs });
    if (ignoreErrors) promise = promise.catch(() => {});
    if (!parallel) {
      await promise;
    } else {
      allPromises.push(promise);
    }
  }

  // If parallel logs are enabled, we have to manually exit
  if (parallel && parallelLogs) {
    await Promise.all(allPromises);
    process.exit(0);
  }
};

export default run;
