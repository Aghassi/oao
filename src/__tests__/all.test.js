/* eslint-env jest */
/* eslint-disable global-require */

import all from '../all';

jest.mock('../utils/shell');

describe('ALL command', () => {
  beforeEach(() => { jest.resetAllMocks(); });

  it('executes the specified command on all sub-packages', async () => {
    const helpers = require('../utils/shell');
    await all('ls', { src: 'test/fixtures/packages/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
