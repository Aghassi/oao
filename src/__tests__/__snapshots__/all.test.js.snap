// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ALL command can do it following the dependency tree 1`] = `
Array [
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-b",
      "storySrc": "oao-b",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao",
      "storySrc": "oao",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-c",
      "storySrc": "oao-c",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-d",
      "storySrc": "oao-d",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-priv",
      "storySrc": "oao-priv",
    },
  ],
]
`;

exports[`ALL command executes the specified command on all sub-packages 1`] = `
Array [
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao",
      "storySrc": "oao",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-b",
      "storySrc": "oao-b",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-c",
      "storySrc": "oao-c",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-d",
      "storySrc": "oao-d",
    },
  ],
  Array [
    "ls",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-priv",
      "storySrc": "oao-priv",
    },
  ],
]
`;
